---
title: "Adult brain cells data analysis"
author: "Health Data Science Unit"
date: "11/14/2019"
output: 
    html_document:
            code_fold: 'hide'
---

```{r echo=FALSE}
## set defaults settings for chunks
knitr::opts_chunk$set(message=FALSE, 
                      warning=FALSE)
```

## Loading data into R

```{r}
library(Seurat)
library(dplyr)

## Loading Adult brain cortex chromatin accesibility reads
chrDir <- '../data/chenS2019sameCellsData/GSE126074_AdBrainCortex_SNAREseq_chromatin/'
adBrchr <- Read10X(data.dir = chrDir, 
                    gene.column = 1      ## necessary parameter, otherwise runs error
                    )
adBrchr.Seu <- CreateSeuratObject(counts = adBrchr,
                                   project = 'ad_chr', 
                                   min.cells = 1,
                                   min.features = 1)
```

## Processing gene expression

```{r}
seurat_pipe_10x <- function(
        data_dir='.',
        gene.column = 1,
        projectName = 'sc_data_analysis',
        min.cells = 1,
        min.features = 1,
        n_features = 1000,
        annotations = FALSE,
        pca_dims = 15) {
        cat('Loading data\n')
        counts_10x <- Read10X(data.dir = data_dir, 
                            ## necessary parameter, otherwise runs error
                            gene.column = gene.column)
        seurat_pipe <- CreateSeuratObject(counts = counts_10x,
                                   project = projectName, 
                                   min.cells = min.cells,
                                   min.features = min.features)
        
        cat('Normalizing and finding variable features\n')
        seurat_pipe <- NormalizeData(seurat_pipe) %>%
                         FindVariableFeatures(selection.method = 'vst',
                                       nfeatures = n_features)
        
        cat('Adding annotations\n')
        if ( class(annotations) %in% c('factor', 'character') &
             length(annotations) == length(colnames(seurat_pipe))
             ) {
                seurat_pipe$'annotations' <- annotations
        }
        
        cat('Scaling and projection\n')
        seurat_pipe <- ScaleData(seurat_pipe, 
                                verbose = FALSE) %>% 
                           RunPCA(npcs = pca_dims, 
                                  verbose = FALSE) %>%
                           RunUMAP(reduction = "pca", 
                                   dims = 1:15)
        
        return(seurat_pipe)
}

## setting pathway
dir_10x <- '../data/chenS2019sameCellsData/GSE126074_AdBrainCortex_SNAREseq_cDNA/'
## Loading annotations
metadata  <- readRDS("AdBrainCortex_SNAREseq_metadata.rds")
adBr_mRNA <- seurat_pipe_10x(
        dir_10x,
        projectName = 'mRNA',
        annotations = metadata$Ident
)
```

```{r}
DimPlot(adBr_mRNA, reduction = 'umap', group.by = 'annotations', label = TRUE) +
        NoLegend()
```

```{r}
adBrchr_sub <- subset(adBrchr.Seu, features = 1:50)
adBrchr_sub
activity.matrix <- CreateGeneActivityMatrix(
        peak.matrix = GetAssayData(adBrchr.Seu, slot='counts'), 
                annotation.file = "../data/Mus_musculus.GRCm38.98.gtf.gz", 
        seq.levels = c(1:19, "X", "Y"), 
        upstream = 2000, 
        verbose = TRUE
)
```





