---
title: "Clustering brain single cells data"
author: "Health Data Science Unit"
date: "11/6/2019"
output: 
    html_document:
            code_fold: 'hide'
---

```{r echo=FALSE}
## set defaults settings for chunks
knitr::opts_chunk$set(message=FALSE, 
                      warning=FALSE,
                      cache = TRUE)
```

## 1. Loading data into R

```{r}
library(Seurat)
library(dplyr)

## Embryonic cells 
## GSE126074_P0_BrainCortex mRNA
cDNADir <- '../data/chenS2019sameCellsData/GSE126074_P0_BrainCortex_SNAREseq_cDNA/'
p0cDNA <- Read10X(
        data.dir = cDNADir, 
        gene.column = 1      ## necessary parameter, otherwise runs error
)
p0cDNA@Dimnames[[1]] <- toupper(p0cDNA@Dimnames[[1]])
p0cDNA.Seu <- CreateSeuratObject(
        counts = p0cDNA,
        project = 'sameCellsData', 
        min.cells = 1,
        min.features = 1
)

p0chrDir <- '../data/chenS2019sameCellsData/GSE126074_P0_BrainCortex_SNAREseq_chromatin/'
p0chr <- Read10X(
        data.dir = p0chrDir,
        gene.column = 1      ## necessary parameter, otherwise runs error
)
p0chr.Seu <- CreateSeuratObject(
        counts = p0chr,
        project = 'sameCellsData', 
        min.cells = 1,
        min.features = 1
)
```

## 2. Pre-processing

```{r}
## Preprocessing
p0cells <- list(p0cDNA.Seu, p0chr.Seu)
names(p0cells) <- c('p0_mRNA', 'p0_chr')
for (i in 1:length(p0cells)){
        p0cells[[i]] <- NormalizeData(p0cells[[i]])
        p0cells[[i]] <- FindVariableFeatures(p0cells[[i]], 
                                              selection.method = 'vst',
                                              nfeatures = 1000)
}
```

## 3. UMAP projection of clusters

Cell annotations were kindly provided by Dr. Kun Zhang. 


```{r}
p0cells[[1]] <- p0cells[[1]] %>% 
                  ScaleData(verbose = FALSE) %>% 
                   RunPCA(npcs = 10, verbose = FALSE) %>%
                    RunUMAP(reduction = "pca", dims = 1:9, metric = 'cosine')
```

```{r}
metadata  <- readRDS("P0BrainCortex_SNAREseq_metadata_full.rds")
p0cells[[1]]$'clust_ann' <- metadata$IdentChar
```

## Gene expression clustering of neonatal cells

<center>
```{r fig.cap='**UMAP projection of the clusters obtained using KNN graph method**.'}
DimPlot(p0cells[[1]], 
        reduction = 'umap', 
        label = TRUE, 
        group.by = 'clust_ann') + 
            NoLegend()
```
</center>


## Information transfer

```{r}
activity.matrix <- CreateGeneActivityMatrix(
        peak.matrix = GetAssayData(p0chr.Seu, slot='counts'), 
        annotation.file = "../data/Homo_sapiens.GRCh37.87.gtf", 
        seq.levels = c(1:22, "X", "Y"), 
        upstream = 2000, 
        verbose = TRUE
)
```


## Processing  gene activity matrix

```{r}
## creation of the Seurat object containing scATAC counts
p0_atac <- CreateSeuratObject(
        counts = GetAssayData(p0chr.Seu, slot = 'counts'),
        assay = 'ATAC',
        project = 'multilayer_integration'
)
## Adding estimated gene activity matrix
p0_atac[['activity']] <- CreateAssayObject(
        counts = activity.matrix
)

```


```{r}
DefaultAssay(p0_atac) <- 'activity'
p0_atac <- FindVariableFeatures(p0_atac) %>%
              NormalizeData() %>%
                ScaleData()
```

## Processing ATAC data

```{r}
DefaultAssay(p0_atac) <- 'ATAC'
VariableFeatures(p0_atac) <- names(which(Matrix::rowSums(p0_atac) > 100))
p0_atac <- RunLSI(p0_atac, n = 50, scale.max = NULL)
p0_atac <- RunUMAP(p0_atac, reduction = "lsi", dims = 1:50)
```

```{r}
DefaultAssay(p0_atac) <- 'activity'
anchors <- FindTransferAnchors(
        reference = p0cells[[1]],
        query = p0_atac, 
        reduction = 'cca'
)
```

```{r}
preds <- TransferData(
        anchorset = anchors,
        refdata = p0cells[[1]]$clust_ann,
        weight.reduction = p0_atac[['lsi']]
)
```


```{r}
## adding metadata to ATAC data from gene expression labels

## getting the order in which cells appears in ATAC data
ord <- match(colnames(p0_atac), 
         colnames(p0cells[[1]])) 
p0_atac[['gene_exp']] <- p0cells[[1]]$clust_ann[ord] 

```

## UMAP projection of scATAC data 

```{r fig.cap='**UMAP projection of ATAC data in neonatal cells tagged using the labels from gene expression clusters manually curated.** '}
#p0_atac[['preds']] <- preds$predicted.id
DimPlot(p0_atac, 
        reduction = 'umap', 
        label = TRUE, 
        group.by = 'gene_exp') + NoLegend()
```

## Transfer learning from gene expression to ATAC 

```{r fig.cap='**UMAP representation of ATAC data labeled by the transferred predicted ids from mRNA dataset.**.'}
p0_atac[['preds']] <- preds$predicted.id
DimPlot(p0_atac, 
        reduction = 'umap', 
        group.by = 'preds')
```

```{r}
p0_atac <- ScaleData(p0_atac, 
                        verbose = FALSE)
p0_atac <- RunPCA(p0_atac, 
                     npcs = 30, 
                     verbose = FALSE)

```

```{r}
p0_atac <- FindNeighbors(p0_atac, dims = 1:15, verbose = FALSE)
p0_atac <- FindClusters(p0_atac, resolution= 1, verbose = FALSE)
```

```{r}
DimPlot(p0_atac,
        group.by = 'seurat_clusters',
        reduction = 'umap',
        label = TRUE) + NoLegend()
```

