---
title: "Heterogeneity in HCV response at the single cell level"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
    html_document:
            code_folding: hide
---

```{r setup, include=FALSE}
library(Matrix)
library(dplyr)
library(uwot)
library(ggplot2)
source("http://bioconductor.org/biocLite.R")
biocLite("GEOquery")
library(GEOquery)
library(Seurat)

knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE,
                      cache = TRUE,
                      fig.width = 10,
                      fig.height = 7, 
                      echo = FALSE)
```

## Hepatitis virus data

Eltahla A A et al, 2016 data of Ag-specific T cells from peripheral blood cells in patients who cleared HCV infection.

```{r data_eltahla_loading, echo=TRUE}
## Raw counts files were downloaded here:
## https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-4850/downloads?geneId=ENSG00000107779

## loading data from Eltahla AA et al, 2016
## Counts matrix
rna <- readMM('../data/hcv/eltahla2016/E-MTAB-4850.aggregated_filtered_counts.mtx')
col_names <- readLines('../data/hcv/eltahla2016/E-MTAB-4850.aggregated_filtered_counts.mtx_cols') 
row_names <- read.table('../data/hcv/eltahla2016/E-MTAB-4850.aggregated_filtered_counts.mtx_rows')
rna <- as.matrix(rna)
colnames(rna) <- col_names
rownames(rna) <- row_names$V1
## Metadata
metadata <- read.table('../data/hcv/eltahla2016/E-MTAB-4850.sdrf.txt', 
                       sep = '\t', header = TRUE, stringsAsFactors = FALSE)
colnames(metadata) <- gsub('\\.', '', colnames(metadata))
metadata.s <- unique(select(metadata, 
                            CommentBioSD_SAMPLE, 
                            SourceName,
                            Characteristicscelltype))
```


```{r umap_vis}
rna_umap <- umap(t(rna), metric = 'cosine')
colnames(rna_umap) <- c('UMAP1', 'UMAP2')
rna_umap.df <- as.data.frame(rna_umap)
rna_umap.df <- mutate(rna_umap.df, id=col_names)
## length(intersect(metadata.s$CommentBioSD_SAMPLE, col_names)) ## check metatata intersection with colnames is unique
metadata.s <- rename(metadata.s, id=CommentBioSD_SAMPLE)
rna_umap.df <- merge(rna_umap.df, metadata.s)
rna_umap.df <- mutate(rna_umap.df, origin= gsub('[0-9]', '', SourceName))
rna_umap.df <- mutate(rna_umap.df, origin= trimws(origin))
ggplot(data = rna_umap.df) +
        geom_point(aes(UMAP1, UMAP2, colour=origin, size=4))
```

```{r rosemberg_data, eval=FALSE}
url_rosenmberg <- 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE119117&format=file&file=GSE119117%5Fcounts%5Fmatrix%2Etxt%2Egz'
if ( !file.exists('../data/hcv/rosemberg2018/rosemberg.gz')) {
       download.file(url = url_rosenmberg, 
                     destfile = '~/sc/singleCellDeconvolution//data/hcv/rosemberg2018/rosemberg.gz')       
}
bulk <- read.table('../data/hcv/rosemberg2018/rosemberg.gz')

## metadata 
## url https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA488222&o=acc_s%3Aa
rosemberg_metadata <- read.csv('../data/hcv/rosemberg2018/rosemberg2018_metadata.csv')
rosemberg_metadata.s <- unique(select(rosemberg_metadata, Sample.Name, time_point))
series_matrix <- getGEO(filename="~/Downloads/GSE119117_series_matrix.txt")
sm <- data.frame(geo_acc=series_matrix$geo_accession, sample=series_matrix$`time point:ch1`,
                 id=series_matrix$title)
sm <- mutate(sm, id=gsub('-', '.', id))
```

```{r eval=FALSE}
bulk_umap <- umap(t(bulk), metric = 'cosine')
bulk_umap.df <- as.data.frame(bulk_umap)
bulk_umap.df <- mutate(bulk_umap.df, id=colnames(bulk))
bulk_umap.df <- merge(bulk_umap.df, sm)
```

```{r eval=FALSE}
## Downloading Satpathy AT et al, 2019 data
barcodes_url <- 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129785&format=file&file=GSE129785%5FscATAC%2DTME%2DTCells%2Ecell%5Fbarcodes%2Etxt%2Egz'
matrix_url <- 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129785&format=file&file=GSE129785%5FscATAC%2DTME%2DTCells%2Emtx%2Egz'
peaks_url <- 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129785&format=file&file=GSE129785%5FscATAC%2DTME%2DTCells%2Epeaks%2Etxt%2Egz'

## dowloading data
if ( ! file.exists('../data/hcv/satpathy2016/barcodes.tsv')) {
      download.file(barcodes_url, destfile = '../data/hcv/satpathy2016/barcodes.tsv.gz')
      gunzip('../data/hcv/satpathy2016/barcodes.tsv.gz', 
             destname = '../data/hcv/satpathy2016/barcodes.tsv',
             remove = FALSE)  
}
if ( ! file.exists('../data/hcv/satpathy2016/matrix.mtx')){
        download.file(matrix_url, destfile = '../data/hcv/satpathy2016/matrix.mtx.gz')
        gunzip('../data/hcv/satpathy2016/matrix.mtx.gz', 
               destname = '../data/hcv/satpathy2016/matrix.mtx', 
               remove = FALSE)  
}
if ( ! file.exists('../data/hcv/satpathy2016/genes.tsv')){
        download.file(peaks_url, destfile = '../data/hcv/satpathy2016/genes.tsv.gz')
        gunzip('../data/hcv/satpathy2016/genes.tsv.gz', 
               destname = '../data/hcv/satpathy2016/genes.tsv',
               remove = FALSE) 
}


## loading data
barcodes <- read.table('../data/hcv/satpathy2016/barcodes.tsv', 
                       header = TRUE) %>% select(Barcodes) 
barcodes <- as.vector(barcodes$Barcodes)


#counts <- Read10X('../data/hcv/satpathy2016/', gene.column = 1)
counts <- readMM('../data/hcv/satpathy2016/matrix.mtx')
counts.s <- counts[, samps]
peaks <- read.table('../data/hcv/satpathy2016/genes.tsv', header = TRUE)

## subsampling
samps <- sample(1:length(barcodes), 1000)
barcodes.s <- barcodes[samps]
counts.s@Dimnames <- list(peaks, barcodes.s)
atac.seu <- CreateSeuratObject(counts = counts.s, project = 'atac')
```

```{r rna_processing}
rna.seu <- CreateSeuratObject(counts = rna, project = 'rna')

rna.seu <- NormalizeData(rna.seu, verbose = FALSE)
rna.seu <- FindVariableFeatures(rna.seu, selection.method = 'vst', nfeatures = 5000)
rna.seu <- ScaleData(rna.seu, verbose = FALSE)
rna.seu <- RunPCA(rna.seu, npcs = 30, verbose = FALSE)
rna.seu <- RunUMAP(rna.seu, reduction = "pca", dims = 1:30)
rna.seu <- FindNeighbors(rna.seu, dims = 1:30, verbose = FALSE)
rna.seu <- FindClusters(rna.seu, resolution= 0.5, verbose = FALSE)
```


```{r atac_processing}
atac.seu <- NormalizeData(atac.seu, verbose = FALSE)
atac.seu <- FindVariableFeatures(atac.seu, selection.method = 'vst', nfeatures = 10000)

activity.matrix <- CreateGeneActivityMatrix(
        peak.matrix = GetAssayData(atac.seu, slot='counts'), 
                annotation.file = "../data/Homo_sapiens.GRCh38.98.gtf.gz", 
        seq.levels = c(1:22, "X", "Y"), 
        upstream = 2000, 
        verbose = TRUE
)
#saveRDS(activity.matrix, '../data/hcv/activity_matrix_sathpathy2019.RDS')
```






