---
title: "Seurat benchmark"
author: "Health Data Science Unit"
date: "10/31/2019"
output: 
   html_document:
           code_folding: 'hide'
---


```{r echo=FALSE}
library(sctransform)

## set defaults settings for chunks
knitr::opts_chunk$set(cache=TRUE, 
                      message=FALSE, 
                      warning=FALSE)
```

Our goal is to test the performance of Seurat information transfer by using single cells
data from SNARE sequencing that allows to quantify cDNA along with chromatin accesibility
reads in the same cells.

## Data

The following cell lines were profiled:

* BJ

* H1

* K562

* GM12878


## Analysis strategy

* Clustering of cells to identify the cell subtypes described above.

* Validation of the clustering analysis using known markers.

* Projection of the labels from gene expression to chromatin accesibility data.

* Cross validation using the known cells ids.


```{r}
library(Seurat)

##############################################################################################
## Cells Mixture cDNA
mixcDNA <- read.table('../data/chenS2019sameCellsData/GSE126074_CellLineMixture_SNAREseq_cDNA_counts.tsv',
                      sep = '\t',
                      stringsAsFactors = FALSE) 
mixcDNA.seu <- CreateSeuratObject(counts = mixcDNA, 
                                  assay = 'mixCDNA')

##############################################################################################

## Cells Mixture chromatin
mixchr <- read.table('../data/chenS2019sameCellsData/GSE126074_CellLineMixture_SNAREseq_chromatin_counts.tsv',
                      sep = '\t',
                      stringsAsFactors = FALSE) 
mixchr.Seu <- CreateSeuratObject(counts = mixchr,
                                 assay = 'mixChr')

```

## Clustering mixture cells

First, the cells are clasified using KNN graph clustering in order to obtain cell subtypes which
are going to be projected into chromosome accesibility dataset. We could reproduce the four clusters observed by Chen and collaborators. 

```{r}
## Preprocessing
mixture <- list(mixcDNA.seu, mixchr.Seu)
names(mixture) <- c('mix_mRNA', 'mix_chr')
for (i in 1:length(mixture)){
        mixture[[i]] <- NormalizeData(mixture[[i]])
        mixture[[i]] <- FindVariableFeatures(mixture[[i]], 
                                              selection.method = 'vst',
                                              nfeatures = 1000)
}
```

```{r}
mixture$mix_mRNA <- ScaleData(mixture$mix_mRNA, verbose = FALSE)
mixture[[1]] <- RunPCA(mixture[[1]], npcs = 30, verbose = FALSE)
mixture[[1]] <- RunUMAP(mixture[[1]], reduction = "pca", dims = 1:30)
mixture[[1]] <- FindNeighbors(mixture[[1]], dims = 1:30, verbose = FALSE)
mixture[[1]] <- FindClusters(mixture[[1]], resolution= 0.5, verbose = FALSE)
```

<center>
```{r, fig.cap='**Clustering of the mRNA single cell data**. Cell labels were assigned according to the expression of cell lines markers.'}
## Assignation of cluster names
cell_labs <- plyr::mapvalues(x=mixture[[1]]$seurat_clusters,
                from=c(0,1,2,3),
                to=c('H1', 'BJ', 'K562', 'GM12878'))
mixture[[1]][['cell_labs']] <- cell_labs

``` 
</center>

## Expression of cell lines markers

Next, we wanted to identify clusters by profiling known markers for each cell line.
BJ, GM12878, H1 and K562 express JUN, IRF8, POUF5F1 and GATA1, respectively. 


<center>
```{r, fig.cap='**Evaluation of the expression of markers**. '}
VlnPlot(mixture[[1]], 
        features = c('JUN', 'IRF8','POU5F1', 'GATA1'),
        ncol = 2)
```
</center>

## UMAP projection of identified clusters

Based in the previous inspection of known expression markers for each cell line,
we perfomed UMAP projection and label the clusters with inferred identities as
shown in the next plot.

<center>
```{r fig.cap='**UMAP projection of the clusters obtained using KNN graph method**. Four groups can be observed which were labeled according to the corresponding expression of cell line markers.'}
DimPlot(mixture[[1]], 
        reduction = 'umap', 
        label = TRUE, 
        group.by = 'cell_labs') + 
            NoLegend()
```
</center>

However, from the next visualization it can be seen that the expression of the markers
in each cluster is heterogenous.

<center>
```{r fig.cap='**Expression of known markers in clusters** as visualized using UMAP projection.'}
FeaturePlot(mixture[[1]], 
            features = c('JUN', 'IRF8','POU5F1', 'GATA1'))
```
</center>


## Information transference

```{r}

activity.matrix <- CreateGeneActivityMatrix(peak.matrix = GetAssayData(mixture[[2]], slot='counts'), 
                                            annotation.file = "../data/Homo_sapiens.GRCh37.87.gtf", 
                                            seq.levels = c(1:22, "X", "Y"), 
                                            upstream = 2000, verbose = TRUE)
mixture.atac <- CreateSeuratObject(counts = GetAssayData(mixture[[2]], slot='counts'),
                                   assay = 'ATAC',
                                   project = "ATAC_mixture_projection")
mixture.atac[['ACTIVITY']] <- CreateAssayObject(counts = activity.matrix)
mixture.atac[['tech']] <- 'atac'

DefaultAssay(mixture.atac) <- "ACTIVITY"
mixture.atac <- FindVariableFeatures(mixture.atac)
mixture.atac <- NormalizeData(mixture.atac)
mixture.atac <- ScaleData(mixture.atac)

DefaultAssay(mixture.atac) <- "ATAC"
VariableFeatures(mixture.atac) <- names(which(Matrix::rowSums(mixture.atac) > 100))
mixture.atac <- RunLSI(mixture.atac, n = 50, scale.max = NULL)
mixture.atac <- RunUMAP(mixture.atac, reduction = "lsi", dims = 1:50)

## needs to find out scATAC Seurat input object

#mixture[[1]] <- ScaleData(mixture[[1]])
#mixture[[1]] <- RunLSI(mixture[[1]])
#mixture[[1]] <- RunUMAP(mixture[[1]], reduction = 'lsi', dims = 1:30)

#mixture[[1]] <- SCTransform(mixture[[1]], assay = 'mixCDNA')
mixture.atac <- SCTransform(mixture.atac, assay = 'ACTIVITY')

mixcDNA.seu <- FindVariableFeatures(mixcDNA.seu,
                                    selection.method = 'vst') 
mixcDNA.seu <- NormalizeData(mixcDNA.seu)
mixcDNA.seu <- SCTransform(mixcDNA.seu, assay = 'mixCDNA')

mixcDNA.seu <- SetAssayData(object = mixcDNA.seu,
                            slot = 'assaysS', 
                            new.data = mixcDNA.seu@assays$SCT,
                            assay = 'RNA')

anchors <- FindTransferAnchors(reference = mixcDNA.seu,
                               query = mixture.atac,
                               reference.assay = 'SCT',
                               query.assay = 'SCT') 


##str(mixture[[2]])
```

TODO:

* Transference of the data labels from gene expression data to chromation accesibility quantifications using Seurat.

* Evaluate the match between transfered data with known labels.