---
title: "SNARE-seq Data Integration"
author: "Health Data Science Unit"
date: "10/29/2019"
output: 
  html_document:
     code_folding: "hide"
---


```{r, echo=FALSE}
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE,
                      cache = TRUE)
```

Here, we present a pipeline for integration and analysis of the single cells data on SNARE-seq from 
[Chen S et al, 2019](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126074) of 
neonatal and adult mouse cerebral cortices. SNARE-seq allows to jointly quantifiy gene 
expression along with chromatin accesibility in the same cells.  

### Analysis strategy

* 1. Integration of adult and neonatal mRNA-seq data.

* 2. Projection of the integrated set of chromosome accesibility data.

* 3. Cross validation of the predicted against barcodes IDs cell identities. 

## Loading the data into R

Data is available in the form of cellRanger 10x Chromium output and can be imported
in a straightforward manner. Display the **code** tab for details.

```{r}
## Dependencies
library(Seurat)
library(dplyr)
library(ggplot2)

#############################################################################################

## Loading data into R

#########################################################################################
## Adult brain cells

## Loading Adult brain cortex mRNA quantified transcripts
cDNADir <- '../data/chenS2019sameCellsData/GSE126074_AdBrainCortex_SNAREseq_cDNA/'
adBrcDNA <- Read10X(data.dir = cDNADir, 
                    gene.column = 1      ## necessary parameter, otherwise runs error
                    )
adBrcDNA.Seu <- CreateSeuratObject(counts = adBrcDNA,
                                   project = 'sameCellsData', 
                                   min.cells = 10,
                                   min.features = 500)

## Embryonic cells 
## GSE126074_P0_BrainCortex mRNA
cDNADir <- '../data/chenS2019sameCellsData/GSE126074_P0_BrainCortex_SNAREseq_cDNA/'
p0cDNA <- Read10X(data.dir = cDNADir, 
                            gene.column = 1      ## necessary parameter, otherwise runs error
        )
p0cDNA.Seu <- CreateSeuratObject(counts = p0cDNA,
                                 project = 'sameCellsData', 
                                 min.cells = 10,
                                 min.features = 500)

## Loading Adult brain cortex chromatin accesibility reads
chrDir <- '../data/chenS2019sameCellsData/GSE126074_AdBrainCortex_SNAREseq_chromatin/'
adBrchr <- Read10X(data.dir = chrDir, 
                    gene.column = 1      ## necessary parameter, otherwise runs error
                    )
adBrchr.Seu <- CreateSeuratObject(counts = adBrchr,
                                   project = 'sameCellsData', 
                                   min.cells = 10,
                                   min.features = 500)

## Loading Adult brain cortex chromatin accesibility reads
p0chrDir <- '../data/chenS2019sameCellsData/GSE126074_P0_BrainCortex_SNAREseq_chromatin/'
p0Brchr <- Read10X(data.dir = p0chrDir, 
                    gene.column = 1      ## necessary parameter, otherwise runs error
                    )
p0Brchr.Seu <- CreateSeuratObject(counts = p0Brchr,
                                   project = 'sameCellsData', 
                                   min.cells = 10,
                                   min.features = 500)

Idents(adBrcDNA.Seu) <- rep('ad_mRNA', nrow(adBrcDNA.Seu))
Idents(p0cDNA.Seu) <- rep('p0_mRNA', nrow(p0cDNA.Seu))
```

## Data preprocessing

Preprocessing comprises the following steps:

* 1. Normalization

* 2. Features selection

```{r}
## Preprocessing
brain <- list(adBrcDNA.Seu, p0cDNA.Seu)
names(brain) <- c('ad_mRNA', 'p0_mRNA')
for (i in 1:length(brain)){
        brain[[i]] <- NormalizeData(brain[[i]])
        brain[[i]] <- FindVariableFeatures(brain[[i]], 
                                              selection.method = 'vst',
                                              nfeatures = 2000)
}
```

## Data integration

Next, we integrate gene expression counts for adult and neonatal brain cells. From the next figure, it can be seen UMAP 
projection of integrated data sets. Notice, that the different assay groups are clustered together. 

```{r}
## data integration
brain.anchors <- FindIntegrationAnchors(object.list = brain, 
                                           dims = 1:30)
brain.integrated <- IntegrateData(anchorset = brain.anchors,
                                     dims = 1:30)


# Visualisation of the integrated data
DefaultAssay(brain.integrated) <- "integrated"
brain.integrated <- ScaleData(brain.integrated, verbose = FALSE)
brain.integrated <- RunPCA(brain.integrated, npcs = 30, verbose = FALSE)
brain.integrated <- RunUMAP(brain.integrated, reduction = "pca", dims = 1:30)
```

<center>
```{r, fig.cap='**Data integration of gene expression in adult and neonatal brain single cells data.** Neonatal (p0) and adult (ad) cells are shown in blue and salmon, respectively.'}
DimPlot(brain.integrated, reduction = "umap", group.by = "ident")
```
</center>

## Cells clustering

Then, a KNN graph based algorithm is used to cluster gene expression patterns. A total of 24 clusters were found using a high value for the `resolution` parameter (=2).

```{r}
brain.integrated <- FindNeighbors(brain.integrated, dims = 1:30, verbose = FALSE)
brain.integrated <- FindClusters(brain.integrated, resolution= 2, verbose = FALSE)
```

<center>
```{r}
DimPlot(brain.integrated, reduction = 'umap', label = TRUE) + NoLegend()
```     
</center>

## Expression of known brain tissue markers

We evaluated the expression of known markers expressed in brain tissue cells. Unc5d and Gadd45g are marker for cycling cells associated to survival mechanisms, whilst Foxp1 and Cux1 are upregulated in upper-layer neurones. 

<center>
```{r, fig.cap='**Gene expression of brain markers** in the integrated data of neonatal and adult single cells.'}
FeaturePlot(brain.integrated, features = c('Unc5d', 'Gadd45g','Foxp1', 'Cux1', 'Hmgm2'))
```
</center>

```{r}
markers <- FindAllMarkers(brain.integrated, 
                          only.pos = TRUE, 
                          min.pct = .10, 
                          logfc.threshold = 0.25)
```

```{r}
top.markers <- markers %>% 
        group_by(cluster) %>% 
        top_n(n = 1, wt = avg_logFC) 
```

```{r}
DoHeatmap(brain.integrated, features = top.markers$gene) + NoLegend()
```

https://static-content.springer.com/esm/art%3A10.1038%2Fs41587-019-0290-0/MediaObjects/41587_2019_290_MOESM3_ESM.xlsx

**TODO:**

* Identification of the clusters as cell tissue subtypes.

* Transference of the clusters labels information from gene expression to chromatin accesiblity data.

* UMAP projection of chromatin accesibility data coloured by predicted transferred labels from gene expression.

* UMAP projection of chromatin accesibility data coloured by known cell identities.
